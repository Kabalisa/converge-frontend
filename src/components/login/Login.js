import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Snackbar } from 'react-toolbox';
import { withRouter } from 'react-router-dom';
import '../../../node_modules/react-responsive-carousel/lib/styles/carousel.min.css';
import CustomCarousel from './Carousel';

import ROUTES from '../../utils/routes';
import '../../assets/styles/login.scss';
import { decodeTokenAndGetUserData, getToken } from '../../utils/Cookie';
import Constants from '../../utils/Constants';
import {
  parseQueryString,
  getItemFromLocalStorage,
  saveItemInLocalStorage,
} from '../../utils/Utilities';
import logo from '../../assets/images/converge-logo.svg';
import Button from '../commons/Button';
import LoginModal from './LoginModal';

// destructor constants to be used
const {
  MRM_TOKEN,
  MESSAGES: { defaultLoginError },
} = Constants;

/**
 * The login component.
 *
 * @extends Component
 *
 * @returns {JSX}
 */
export class Login extends Component {
  static propTypes = {
    location: PropTypes.shape({
      pathname: PropTypes.string.isRequired,
      search: PropTypes.string,
      state: PropTypes.object,
    }).isRequired,
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }).isRequired,
  };

  state = {
    loginError: null,
    openModal: false,
    closeModal: true,
    slideAutoplay: true,
  };

  /**
   * Gets the derived state from the props
   *
   * @param {array} props
   *
   * @returns {void}
   */
  static getDerivedStateFromProps = (props) => {
    const { push } = props.history;
    const token = getItemFromLocalStorage(MRM_TOKEN);
    const activeTab =
      sessionStorage.getItem('activeTopNav') || ROUTES.analytics;
    // this logic has been placed in this lifecycle method to avoid rendering of
    // the login page UI if a user is already loggedin
    // if user is logged in
    if (token) {
      // redirect user to settings page
      push(activeTab);
    }
    // returning null since no state updates
    return null;
  };

  componentDidMount = () => {
    const { push } = this.props.history;

    // parse cookie to see if a user has been succesfully logged in by the Andela login api
    const tokenFromCookies = getToken();
    // decode token to get user data
    const data = decodeTokenAndGetUserData();
    if (data && 'UserInfo' in data) {
      // save token to local storage
      saveItemInLocalStorage(MRM_TOKEN, tokenFromCookies);
      // redirect user to settings page
      push(ROUTES.settings);
      return; // avoid execution of error handling if no errors occur
    }

    // handle login errors from the Andela API if they occur
    const { search = '', state: locationState } = this.props.location;
    const queryParameters = parseQueryString(search) || {};
    if ('error' in queryParameters) {
      this.handleLoginError();
    }
    // handle errors that occur when a user tries to access a protected page
    if (locationState && 'errorMessage' in locationState) {
      this.handleLoginError(locationState.errorMessage);
    }
  };

  /**
   * shows login error
   *
   * @returns {void}
   */
  handleLoginError = (loginError = defaultLoginError) => {
    this.setState({ loginError });
  };

  /**
   * It closes the snack bar and redirects to home page
   *
   * @returns {void}
   */
  handleSnackBarClose = () => {
    this.setState({ loginError: null });
    this.props.history.push(ROUTES.home);
  };

  handleOpenModal = () => {
    this.setState({ openModal: true, closeModal: false, slideAutoplay: false });
  }

  handleCloseModal = () => {
    this.setState({ openModal: false, closeModal: true, slideAutoplay: true });
  }

  render() {
    const { loginError } = this.state;
    const { openModal, closeModal, slideAutoplay } = this.state;
    return (
      <Fragment>
        <LoginModal
          openModal={openModal}
          closeModal={closeModal}
          handleCloseModal={this.handleCloseModal}
        />
        <header className="landing-page-header">
          <div className="header-logo">
            <img src={logo} alt="logo" />
          </div>
          <div className="login-btn">
            {openModal || <Button
              classProp="trigger-login-modal"
              title="Login"
              handleClick={this.handleOpenModal}
            />}
          </div>
        </header>
        <CustomCarousel
          legendPosition={openModal ? 'legend-left' : 'legend-center'}
          autoplay={slideAutoplay}
        />

        {/* display login errors */}
        {loginError && (
          <Snackbar
            type="cancel"
            action="Dismiss"
            label={loginError}
            timeout={Constants.ERROR_TIMEOUT}
            active={Boolean(loginError)}
            onTimeout={this.handleSnackBarClose}
            onClick={this.handleSnackBarClose}
            theme={{ snackbar: 'error-theme', button: 'button-error' }}
          />
        )}
      </Fragment>
    );
  }
}

export default withRouter(Login);
