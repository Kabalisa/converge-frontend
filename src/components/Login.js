import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import ROUTES from '../utils/routes';
import '../assets/styles/login.scss';
import darkTabletImage from '../assets/images/DarkTablet.png';
import loginButton from '../assets/images/Button.png';

/**
 * The login component. Currently only implements the Login  User Interface
 * Login functionality to be Added Later
 * @author [Dennis Jjagwe](https://github.com/dennisja)
 */

class Login extends Component {
  static propTypes = {
    location: PropTypes.shape({ pathname: PropTypes.string.isRequired }),
  };
  componentDidMount() {
    this.applyBodyStyles();
  }
  /**
   * Logs in a user
   */
  loginUser = () => {
    // login logic here
  };
  // Apply body styles on homepage only
  applyBodyStyles = () => {
    const { location } = this.props;
    location.pathname === ROUTES.home && document.body.classList.add('home-bg');
  };

  render() {
    return (
      <Fragment>
        <div>
          <header className="home-header">
            <h1>CONVERGE</h1>
          </header>

          <div id="converge-intro">
            Meet the Meeting Room App
            <br />
            that your meeting room app aspires to be.
          </div>

          <input
            type="image"
            src={loginButton}
            alt="Login With Your Email"
            className="btn-signin"
            onClick={this.loginUser}
          />
        </div>

        <img src={darkTabletImage} alt="Dark Tablet" id="dark-tablet" />
      </Fragment>
    );
  }
}

export default withRouter(Login);
