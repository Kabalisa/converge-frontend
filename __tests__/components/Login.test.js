import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import { shallow, mount } from 'enzyme';

import Login from '../../src/components/Login';

describe('Login component', () => {
  const wrapper = <MemoryRouter keyLength={0}><Login /></MemoryRouter>;
  const mountWrapper = mount(wrapper);
  const shallowWrapper = shallow(wrapper);

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(wrapper, div);
  });

  it('renders correctly', () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('renders the correct heading text', () => {
    expect(mountWrapper.find('header>h1')).toHaveLength(1);
    expect(mountWrapper.find('header>h1').text()).toBe('CONVERGE');
  });

  it('renders the correct mrm introduction text', () => {
    const mrmIntro =
      'Meet the Meeting Room Appthat your meeting room app aspires to be.';
    expect(mountWrapper.find('#converge-intro')).toHaveLength(1);
    expect(mountWrapper.find('#converge-intro').text()).toBe(mrmIntro);
  });

  it('renders the tablet image', () => {
    expect(mountWrapper.find('#dark-tablet img')).toHaveLength(1);
  });

  it('renders login button', () => {
    //   check whether login button exists
    expect(mountWrapper.find('.btn-signin')).toHaveLength(1);
  });
});
