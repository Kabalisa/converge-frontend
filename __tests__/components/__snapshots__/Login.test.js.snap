// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(Login)>
      <Route
        render={[Function]}
      >
        <Login
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <header
            className="home-header"
          >
            <h1>
              CONVERGE
            </h1>
          </header>
          <Component>
            <div
              id="converge-intro"
            >
              Meet the Meeting Room App
              <br />
              that your meeting room app aspires to be.
            </div>
          </Component>
          <LoginButton>
            <a
              href="https://api-staging.andela.com/login?redirect_url=http://mrm-dev.andela.com:8080"
            >
              <input
                alt="Login With Your Email"
                className="btn-signin"
                src="test-file-stub"
                type="image"
              />
            </a>
          </LoginButton>
          <img
            alt="Dark Tablet"
            id="dark-tablet"
            src="test-file-stub"
          />
        </Login>
      </Route>
    </withRouter(Login)>
  </Router>
</MemoryRouter>
`;

exports[`Login component should call componentWillUnmount when unmounted 1`] = `
<Login
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "pathname": "/",
      "state": Object {
        "errorMesage": "Something Went Wrong",
      },
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="https://api-staging.andela.com/login?redirect_url=http://mrm-dev.andela.com:8080"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="test-file-stub"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="test-file-stub"
  />
</Login>
`;

exports[`Login component should display an error when login error occurs and close snackbar on time out 1`] = `
<Login
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "pathname": "/",
      "search": "?error=Something%20Went%20Wrong",
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="https://api-staging.andela.com/login?redirect_url=http://mrm-dev.andela.com:8080"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="test-file-stub"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="test-file-stub"
  />
  <ThemedActivableRenderer
    action="Dismiss"
    active={true}
    composeTheme="deeply"
    delay={500}
    label="Please Login With Your Andela Email"
    mapThemrProps={[Function]}
    onClick={[Function]}
    onTimeout={[Function]}
    theme={
      Object {
        "button": "button-error",
        "snackbar": "error",
      }
    }
    timeout={6000}
    type="cancel"
  >
    <ActivableRenderer
      action="Dismiss"
      active={true}
      delay={500}
      label="Please Login With Your Andela Email"
      onClick={[Function]}
      onTimeout={[Function]}
      theme={
        Object {
          "button": "button-error",
          "snackbar": "error",
        }
      }
      timeout={6000}
      type="cancel"
    >
      <Snackbar
        action="Dismiss"
        active={true}
        label="Please Login With Your Andela Email"
        onClick={[Function]}
        onTimeout={[Function]}
        theme={
          Object {
            "button": "button-error",
            "snackbar": "error",
          }
        }
        timeout={6000}
        type="cancel"
      >
        <Portal
          className=""
        />
      </Snackbar>
    </ActivableRenderer>
  </ThemedActivableRenderer>
</Login>
`;

exports[`Login component should display an error when login error occurs and close snackbar on time out 2`] = `
<Login
  history={
    Object {
      "push": [MockFunction] {
        "calls": Array [
          Array [
            "/",
          ],
        ],
      },
    }
  }
  location={
    Object {
      "pathname": "/",
      "search": "?error=Something%20Went%20Wrong",
    }
  }
>
  <header
    className="home-header"
  >
    <h1>
      CONVERGE
    </h1>
  </header>
  <Component>
    <div
      id="converge-intro"
    >
      Meet the Meeting Room App
      <br />
      that your meeting room app aspires to be.
    </div>
  </Component>
  <LoginButton>
    <a
      href="https://api-staging.andela.com/login?redirect_url=http://mrm-dev.andela.com:8080"
    >
      <input
        alt="Login With Your Email"
        className="btn-signin"
        src="test-file-stub"
        type="image"
      />
    </a>
  </LoginButton>
  <img
    alt="Dark Tablet"
    id="dark-tablet"
    src="test-file-stub"
  />
</Login>
`;
